src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ReflectMethods.java

```diff
@@ -1492,360 +1492,226 @@ public void visitIf(JCTree.JCIf tree) {
-                    pushBody(headCl, FunctionType.VOID);
+                append(CoreOp._yield(localResult));
+                body = stack.body;
 
-                    append(CoreOp._yield());
-                    defaultLabel = stack.body;
+                // Pop labels
+                popBody();
+            } else if (headCl instanceof JCTree.JCDefaultCaseLabel) {
+                // @@@ Do we need to model the default label body?
+                pushBody(headCl, FunctionType.VOID);
 
-                    // Pop label
-                    popBody();
-                } else {
-                    throw unsupported(tree);
-                }
+                append(CoreOp._yield());
+                body = stack.body;
 
-                // Statements body
-                switch (c.caseKind) {
-                    case RULE -> {
-                        pushBody(c.body, actionType);
-                        if (c.body instanceof JCTree.JCBlock b) {
-                            toValue(b);
-                            if (!(b.stats.last() instanceof JCTree.JCBreak)) {
-                                append(CoreOp._yield()); // @@@ _break is also an option
-                            }
-                        }
-                        else if (c.body instanceof JCTree.JCStatement s) {
-                            toValue(s);
-                            if (!(s instanceof JCTree.JCThrow)) {
-                                append(CoreOp._yield());
-                            }
-                        }
+                // Pop label
+                popBody();
+            } else {
+                throw unsupported(tree);
+            }
 
-                        if (headCl instanceof JCTree.JCDefaultCaseLabel) {
-                            defaultStatements = stack.body;
-                        } else {
-                            bodies.add(stack.body);
-                        }
+            return body;
+        }
 
-                        // Pop block
-                        popBody();
-                    }
-                    case STATEMENT -> {
-                        // @@@ Avoid nesting for a single block? Goes against "say what you see"
-                        // boolean oneBlock = c.stats.size() == 1 && c.stats.head instanceof JCBlock;
-                        pushBody(c, actionType);
+        private Body.Builder visitCaseBody(JCTree.JCCase c, FunctionType caseBodyType) {
 
-                        scan(c.stats);
+            Body.Builder body = null;
+            Type yieldType = typeElementToType(caseBodyType.returnType());
 
-                        appendTerminating(c.completesNormally ?
-                                headCl instanceof JCTree.JCDefaultCaseLabel ? CoreOp::_yield : ExtendedOp::switchFallthroughOp
-                                : CoreOp::unreachable);
+            JCTree.JCCaseLabel headCl = c.labels.head;
+            switch (c.caseKind) {
```

Comment:
Suggestion, up to you. You can make this a switch expression yielding the body builder and directly return the result
```
return switch (c.caseKind) {
...

  try {
    yield stack.body;
  } finally {
    popBody();
  }
...

```